resim/msg/default_converter_plugin.cc

```diff
@@ -0,0 +1,174 @@
+// Copyright 2023 ReSim, Inc.
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+#include <cstdint>
+#include <cstring>
+#include <functional>
+#include <rclcpp/serialization.hpp>
+#include <rclcpp/serialized_message.hpp>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+#include "resim/assert/assert.hh"
+#include "resim/msg/converter_plugin_types.h"
+#include "resim/msg/detection_from_ros2.hh"
+#include "resim/msg/header_from_ros2.hh"
+#include "resim/msg/navsat_from_ros2.hh"
+#include "resim/msg/odometry_from_ros2.hh"
+#include "resim/msg/oriented_box_from_ros2.hh"
+#include "resim/msg/pose_from_ros2.hh"
+#include "resim/msg/time_from_ros2.hh"
+#include "resim/msg/transform_from_ros2.hh"
+#include "resim/utils/proto/dependency_file_descriptor_set.hh"
+
+namespace resim::msg {
+
+namespace {
+
+// A helper struct for holding a Converter and SchemaGetter together in our map
+// of ROS2 types to converters (see below)
+struct ConverterFunctions {
+  using Converter = std::function<
+      void(const rcutils_uint8_array_t *, rcutils_uint8_array_t *)>;
+  using SchemaGetter = std::function<void(ReSimConverterSchemaInfo *)>;
+
+  Converter converter;
+  SchemaGetter schema_getter;
+};
+
+// This is the primary function used to convert a given serialized Ros2Type to
+// a serialized ReSimType.
+template <typename Ros2Type>
+void convert_message(
+    const rcutils_uint8_array_t *const ros2_message,
+    rcutils_uint8_array_t *const resim_message) {
+  REASSERT(ros2_message != nullptr);
+  REASSERT(ros2_message->buffer != nullptr);
+  REASSERT(resim_message != nullptr);
+
+  rclcpp::Serialization<Ros2Type> serialization;
+
+  Ros2Type deserialized;
+  rclcpp::SerializedMessage message{*ros2_message};
+  serialization.deserialize_message(&message, &deserialized);
+  const auto converted_message = convert_from_ros2(deserialized);
+
+  const std::size_t converted_size = converted_message.ByteSizeLong();
+
+  rcl_serialized_message_t &result = *resim_message;
+  result.allocator = ros2_message->allocator;
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+  result.buffer = reinterpret_cast<uint8_t *>(
```

Comment:
I should have commented this on previous PR, but we should make clear that this allocate is allowed and will be deallocated by `ConverterPlugin::convert(...)`

Reply:
Good call. I've added some comments on this to converter_plugin.hh